docs/proposals/20210203-unmanaged-infrastructure.md

```diff
@@ -0,0 +1,219 @@
+
+#### Provider implementation changes
+
+To enable providers to implement the changes required by this contract, we will provide a new `InfraClusterExternallyManaged` predicate.
+
+This will filter out any InfraCluster resource that has been marked as "externally managed" and will prevent the controller from reconciling the resource.
+
+### Security Model
+
+When externally managed, no additional privileges for a cloud provider need to be given to CAPI other than the required to manage machines.
+
+### Risks and Mitigations
+
+#### What happens when a user coverts an externally managed InfraCluster to a managed InfraCluster?
+
+Whether the flag for being managed is an annotation or field on the InfraCluster, there currently is no immutability support for CRD fields or annotations within the Kubernetes API.
+
+This means that, once a user has created their externally managed InfraCluster, they could at some point, update the annotation/field to make the InfraCluster appear to be managed.
+
+There is no way to predict what would happen in this scenario.
+The InfraCluster controller would start attempting to reconcile infrastructure that it did not create, and therefore, there may be assumptions it makes that mean it cannot manage this infrastructure.
+
+To prevent this, we will have to implement (in the InfraCluster webhook) a means to prevent users converting externally managed InfraClusters into managed InfraClusters.
+
+Note however, converting from managed to externally managed should cause no issues and should be allowed.
+It will be documented as part of the externally managed contract that this is a one way operation.
+
+### Future Work
+
+#### Marking InfraCluster ready manually
+
+The content of this proposal assumes that the management of the external infrastructure is done by some controller which has the ability to set the spec and status of the InfraCluster resource.
+
+In reality, this may not be the case. For example, if the infrastructure was created by an admin using Terraform.
+
+When using a system such as this, a user can copy the details from the infrastructure into an InfraCluster resource and create this manually.
+However, they will not be able to set the InfraCluster to ready as this requires updating the resource status which is difficult when not using a controller.
+
+To allow users to adopt this external management pattern without the need for writing their own controllers or tooling, we will provide a longer term solution that allows a user to indicate that the infrastructure is ready and have the status set appropriately.
+
+The exact mechanism for how this will work is undecided, though the following ideas have been suggested:
+
+- Write a kubectl plugin that allows a user to mark their InfraCluster resources as ready
+
+- Add a secondary annotation to this contract that causes the provider InfraCluster controller to mark resources as ready
```

Comment:
Assuming that this requires all the infrastructure providers to reconcile the externally managed infra clusters, I'm +1 to remove this option from the list

Reply:
ok to keep to as a record, but please add a note reporting the outcome of the discussion.

Reply:
Has this been done?

Reply:
I would also prefer to not implement this option as it's basically additional complexity in ClusterAPI because of a (current) limitation of kubectl which will probably be resolved mid-term: [[WIP] kubectl: support --subresource flag #99556](https://github.com/kubernetes/kubernetes/pull/99556)